File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/ColumnExamples.java
package questpdf.docexamples;

import com.questpdf.fluent.Document;
import com.questpdf.helpers.Colors;
import com.questpdf.helpers.PageSize;
import com.questpdf.infrastructure.IContainer;
import com.questpdf.infrastructure.ImageCompressionQuality;
import com.questpdf.infrastructure.ImageFormat;
import com.questpdf.infrastructure.ImageGenerationSettings;
import org.junit.jupiter.api.Test;

public class ColumnExamples extends DocExample {

    @Test
    public void simpleExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(250f, 0f));
                    page.maxSize(new PageSize(250f, 1000f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.item().background(Colors.Grey.getMedium()).height(50f);
                            column.item().background(Colors.Grey.getLighten1()).height(75f);
                            column.item().background(Colors.Grey.getLighten2()).height(100f);
                        });
                });
            })
            .generateImages(index -> output("column-simple.webp"), settings);
    }

    @Test
    public void spacingExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(250f, 0f));
                    page.maxSize(new PageSize(250f, 1000f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(25f);

                            column.item().background(Colors.Grey.getMedium()).height(50f);
                            column.item().background(Colors.Grey.getLighten1()).height(75f);
                            column.item().background(Colors.Grey.getLighten2()).height(100f);
                        });
                });
            })
            .generateImages(index -> output("column-spacing.webp"), settings);
    }

    @Test
    public void customSpacingExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(250f, 0f));
                    page.maxSize(new PageSize(250f, 1000f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.item().background(Colors.Grey.getDarken1()).height(50f);
                            column.item().height(10f);
                            column.item().background(Colors.Grey.getMedium()).height(50f);
                            column.item().height(20f);
                            column.item().background(Colors.Grey.getLighten1()).height(50f);
                            column.item().height(30f);
                            column.item().background(Colors.Grey.getLighten2()).height(50f);
                        });
                });
            })
            .generateImages(index -> output("column-spacing-custom.webp"), settings);
    }

    @Test
    public void disableUniformItemsWidthExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(400f, 0f));
                    page.maxSize(new PageSize(400f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);
                    page.pageColor(Colors.getWhite());

                    page.content()
                        .column(column -> {
                            column.spacing(15f);

                            labelStyle(column.item())
                                .text("REST API");

                            labelStyle(column.item())
                                .text("Garbage Collection");

                            labelStyle(column.item())
                                .text("Object-Oriented Programming");
                        });
                });
            })
            .generateImages(index -> output("column-uniform-width-disabled.webp"), settings);
    }

    // The C# original routes LabelStyle through the Element(Func<IContainer, IContainer>)
    // overload, which is not bridged; a static helper method is the direct equivalent.
    private static IContainer labelStyle(IContainer container) {
        return container
            .shrinkHorizontal()
            .background(Colors.Grey.getLighten3())
            .cornerRadius(15f)
            .padding(15f);
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/ColumnExamples.kt
package questpdf.docexamples

import com.questpdf.fluent.Document
import com.questpdf.helpers.Colors
import com.questpdf.helpers.PageSize
import com.questpdf.infrastructure.IContainer
import com.questpdf.infrastructure.ImageCompressionQuality
import com.questpdf.infrastructure.ImageFormat
import com.questpdf.infrastructure.ImageGenerationSettings
import org.junit.jupiter.api.Test

class ColumnExamples : DocExample() {

    @Test
    fun simpleExample() {
        Document
            .create {
                page {
                    minSize(PageSize(250f, 0f))
                    maxSize(PageSize(250f, 1000f))
                    margin(25f)

                    content()
                        .column {
                            item().background(Colors.Grey.Medium).height(50f)
                            item().background(Colors.Grey.Lighten1).height(75f)
                            item().background(Colors.Grey.Lighten2).height(100f)
                        }
                }
            }
            .generateImages({ output("column-simple.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun spacingExample() {
        Document
            .create {
                page {
                    minSize(PageSize(250f, 0f))
                    maxSize(PageSize(250f, 1000f))
                    margin(25f)

                    content()
                        .column {
                            spacing(25f)

                            item().background(Colors.Grey.Medium).height(50f)
                            item().background(Colors.Grey.Lighten1).height(75f)
                            item().background(Colors.Grey.Lighten2).height(100f)
                        }
                }
            }
            .generateImages({ output("column-spacing.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun customSpacingExample() {
        Document
            .create {
                page {
                    minSize(PageSize(250f, 0f))
                    maxSize(PageSize(250f, 1000f))
                    margin(25f)

                    content()
                        .column {
                            item().background(Colors.Grey.Darken1).height(50f)
                            item().height(10f)
                            item().background(Colors.Grey.Medium).height(50f)
                            item().height(20f)
                            item().background(Colors.Grey.Lighten1).height(50f)
                            item().height(30f)
                            item().background(Colors.Grey.Lighten2).height(50f)
                        }
                }
            }
            .generateImages({ output("column-spacing-custom.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun disableUniformItemsWidthExample() {
        Document
            .create {
                page {
                    minSize(PageSize(400f, 0f))
                    maxSize(PageSize(400f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)
                    pageColor(Colors.White)

                    content()
                        .column {
                            spacing(15f)

                            // The C# original routes LabelStyle through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; an extension function is the direct equivalent.
                            fun IContainer.labelStyle(): IContainer =
                                shrinkHorizontal()
                                    .background(Colors.Grey.Lighten3)
                                    .cornerRadius(15f)
                                    .padding(15f)

                            item()
                                .labelStyle()
                                .text("REST API")

                            item()
                                .labelStyle()
                                .text("Garbage Collection")

                            item()
                                .labelStyle()
                                .text("Object-Oriented Programming")
                        }
                }
            }
            .generateImages({ output("column-uniform-width-disabled.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }
}
